### ISSUE TITLE
Combat screen, obstacles and their position on the battlefield should be recorded in the scenario save files

### ISSUE THREAD

**Issue by LeHerosInconnu** [ihhub/fheroes2#7146]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

At the start of the scenario, all obstacles and their position on the battlefield should be determined for all squares where a fight can take place.
This should also be recorded in the scenario save files.

Here a screenshot from a saved scenario where the configuration of the obstacles is different.
Note: This scenario comes from another player on a different configuration than the one on which it was loaded.

In fheroes2:
Screenshot from the scenario video:

![Obstacle type position 002](https://user-images.githubusercontent.com/43583994/236864376-7355d47b-a1eb-44ab-8d87-68f1610a4f69.png)

Screenshot from the saved scenario:

![Obstacle type position 001](https://user-images.githubusercontent.com/43583994/236864360-71087eed-67d6-4a87-a314-96a518a6b74a.png)

### Save file

Related issue (with save file [Scared Golems.zip](https://github.com/ihhub/fheroes2/files/11420970/Scared.Golems.zip)): https://github.com/ihhub/fheroes2/issues/7141.

### Additional info

Scénario loaded in version 1.0.3 7084.

**Comment by oleg-derevenetz** [comment]
> At the start of the scenario, all obstacles and their position on the battlefield should be determined for all squares where a fight can take place.

How is this a bug? Currently it works this way: there is a random seed saved in the save file, and obstacle generation is based on this seed. But the implementation of random generator may be different depending on OS/compiler/etc, so there is a guarantee that you will get the same results (same obstacles on the battlefield in this case) on the same machine, but there is no guarantee that you'll get the same results on the different machine. Why do you expect the same results in the first place? No one promises that.

**Comment by oleg-derevenetz** [comment]
> where all battlefields are the same no matter when you've started new walkthrough

Yes, in OG you'll get the same obstacles on the same tile in different playthroughs. But is that a good thing? Currently you'll get the same obstacles on the same tile during the same playthrough (on the same machine), but during the next playthrough it would be different. I believe it's good because it adds variety to the game.

**Comment by oleg-derevenetz** [comment]
> a file which I copied from my other machine had different obstacles

That's because random distributions (such as `std::uniform_int_distribution` from C++ standard library) may differ depending on OS or C++ compiler flavor/version. To get the same results on different machines, we should implement our own home-grown random generator, which will give the same results regardless of all these things, which is currently absolutely not a priority.
